### PR TITLE
Save and restore clock info for MSHV

### PR DESCRIPTION
**Pull request by russell-islam**
Implement get/set_clock for MSHV and make the clock info independent of hypervisors.

### CODE REVIEW

File: hypervisor/src/lib.rs [hunk 1]
```diff
@@ -165,7 +165,7 @@ pub enum ClockData {
     #[cfg(feature = "kvm")]
     Kvm(kvm_bindings::kvm_clock_data),
     #[cfg(feature = "mshv")]
-    Mshv, /* MSHV does not support ClockData yet */
+    Mshv(mshv::MshvClockData), /* MSHV does not support ClockData yet */
```
Comment (NunoDasNeves) [inline hypervisor/src/lib.rs:168]: remove the comment

File: hypervisor/src/mshv/mod.rs [hunk 2]
```diff
@@ -1930,13 +1949,24 @@ impl vm::Vm for MshvVm {
...
+            .map_err(|e| vm::HypervisorVmError::GetClock(e.into()))?;
+        Ok(MshvClockData { time_ref: val }.into())
     }
 
     /// Set guest clock.
     #[cfg(target_arch = "x86_64")]
     fn set_clock(&self, _data: &ClockData) -> vm::Result<()> {
-        Ok(())
+        let data: MshvClockData = (*_data).into();
+        println!("Setting Time ref Info");
```
Comment (liuw) [inline hypervisor/src/mshv/mod.rs:1963]: Remove this println

File: hypervisor/src/mshv/x86_64/mod.rs [hunk 3]
```diff
@@ -43,6 +43,10 @@ pub struct VcpuMshvState {
     pub vp_states: AllVpStateComponents,
 }
 
+#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize)]
+pub struct MshvClockData {
+    pub time_ref: u64,
```
Comment (liuw) [inline hypervisor/src/mshv/x86_64/mod.rs:48]: The is the reference time? If so, name this `ref_time`.

File: hypervisor/src/mshv/mod.rs [hunk 4]
```diff
@@ -1930,13 +1949,24 @@ impl vm::Vm for MshvVm {
...
+            .get_partition_property(hv_partition_property_code_HV_PARTITION_PROPERTY_REFERENCE_TIME)
+            .map_err(|e| vm::HypervisorVmError::GetClock(e.into()))?;
+        Ok(MshvClockData { time_ref: val }.into())
     }
 
     /// Set guest clock.
     #[cfg(target_arch = "x86_64")]
     fn set_clock(&self, _data: &ClockData) -> vm::Result<()> {
-        Ok(())
+        let data: MshvClockData = (*_data).into();
```
Comment (liuw) [inline hypervisor/src/mshv/mod.rs:1962]: `_data` should be changed to `data`.

**Review by liuw — APPROVED** [review]

**Review by NunoDasNeves — APPROVED** [review]

**Comment by russell-islam** [comment]
@likebreath  PTAL here.

**Comment by rbradford** [comment]
@russell-islam You need to rebase your PR.

File: hypervisor/src/mshv/x86_64/mod.rs [hunk 5]
```diff
@@ -43,6 +43,10 @@ pub struct VcpuMshvState {
     pub vp_states: AllVpStateComponents,
 }
 
+#[derive(Clone, Copy, Debug, Default, Serialize, Deserialize)]
+pub struct MshvClockData {
+    pub ref_time: u64,
+}
```
Comment (rbradford) [inline hypervisor/src/mshv/x86_64/mod.rs:49]: nit: I think there should be a whitespace line here.

File: hypervisor/src/mshv/mod.rs [hunk 6]
```diff
@@ -92,6 +92,25 @@ impl From<mshv_user_mem_region> for UserMemoryRegion {
...
+}
+
+#[cfg(target_arch = "x86_64")]
+impl From<ClockData> for MshvClockData {
+    fn from(ms: ClockData) -> Self {
+        match ms {
+            ClockData::Mshv(s) => s,
+            /* Needed in case other hypervisors are enabled */
+            #[allow(unreachable_patterns)]
+            _ => panic!("MSHV clock data is not valid"),
```
Comment (rbradford) [inline hypervisor/src/mshv/mod.rs:109]: You can use `unreachable!()` here.

**Review by likebreath — COMMENTED** [review]
LGTM. Please rebase and address comments above. Thank you.

File: hypervisor/src/mshv/x86_64/mod.rs (hunk 5 above)
Comment (russell-islam) [reply hypervisor/src/mshv/x86_64/mod.rs:49]: Fixed.

File: hypervisor/src/mshv/mod.rs (hunk 6 above)
Comment (russell-islam) [reply hypervisor/src/mshv/mod.rs]: Fixed

**Review by likebreath — APPROVED** [review]

**Pull request merged**